torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -355,16 +372,26 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
         } else if (message.isResponse()) {
           auto id = message.id();
           std::shared_ptr<FutureMessage> fm = nullptr;
+          std::chrono::milliseconds futureStartTime;
           {
             std::lock_guard<std::mutex> lock{futureMutex_};
-            fm = futures_[id];
+            std::tie(fm, futureStartTime) = futures_[id];
           }
           // Not holding lock on markCompleted as this could run callbacks that
           // call agent_->send
           fm->markCompleted(std::move(message));
           {
             std::lock_guard<std::mutex> lock{futureMutex_};
             futures_.erase(id);
+            // look up the corresponding future by its time out and request ID,
+            // and remove it from the timeouts map
+            auto& futuresAtTime = futureTimeouts_[futureStartTime];
+            futuresAtTime.erase(
+                std::find(futuresAtTime.begin(), futuresAtTime.end(), id));
+            if (futuresAtTime.size() == 0) {
+              // remove the key from futureTimeouts_
+              futureTimeouts_.erase(futureStartTime);
```

Comment:
We will add the background thread in a separate PR?

Reply:
yes, i will do it in follow up diffs.